index.js

```diff
@@ -528,80 +574,140 @@ Client.prototype._destroy = function (err, cb) {
+  // conclude.
+  // - intake response success - A successful response from the APM server. This
+  //   is the normal operation case described above.
+  // - gzipStream error - An "error" event on the gzip stream.
+  // - intake request error - An "error" event on the intake HTTP request, e.g.
+  //   ECONNREFUSED or ECONNRESET.
+  // - intakeResTimeout - A timer started *after* we are finished sending data
+  //   to the APM server by which we require a response (including its body). By
+  //   default this is 10s -- a very long time to allow for a slow or far
+  //   apm-server. If we hit this, APM server is problematic anyway, so the
+  //   delay doesn't add to the problems.
+  // - serverTimeout - An idle timeout value (default 30s) set on the socket.
+  //   This is a catch-all fallback for an otherwised wedged connection. If this
+  //   is being hit, there is some major issue in the application (possibly a
+  //   bug in the APM agent).
+  // - process completion - The Client takes pains to always `.unref()` its
+  //   handles to never keep a using process open if it is ready to exit. When
+  //   the process is ready to exit, the following happens:
+  //    - The "beforeExit" handler above will call `client.end()`,
+  //    - which calls `client._ref()` (to *hold the process open* to complete
+  //      this request), then `_chopper.end()` to end the `gzipStream` so
+  //      this request can complete soon.
+  //    - We then expect this request to complete quickly and the process will
+  //      then finish exiting. A subtlety is if the APM server is not responding
+  //      then we'll wait on `intakeResTimeoutOnEnd` (by default 1s).
+  return function makeIntakeRequest (gzipStream, next) {
+    const reqId = crypto.randomBytes(16).toString('hex')
+    const log = client._log.child({ reqId })
+    const startTime = process.hrtime()
+    const timeline = []
+    let bytesWritten = 0
+    let intakeRes
+    let intakeResTimer = null
+    const intakeResTimeout = client._conf.intakeResTimeout
+    const intakeResTimeoutOnEnd = client._conf.intakeResTimeoutOnEnd
+
+    // `_active` is used to coordinate the callback to `client.flush(db)`.
+    client._active = true
 
-    // Mointor streams for errors so that we can make sure to destory the
-    // output stream as soon as that occurs
-    stream.on('error', onerrorproxy)
-    req.on('error', onerrorproxy)
+    // Handle conclusion of this intake request. Each "part" is expected to call
+    // `completePart()` at least once -- multiple calls are okay for cases like
+    // the "error" and "close" events on a stream being called. When a part
+    // errors or all parts are completed, then we can conclude.
+    let concluded = false
+    const completedFromPart = {
+      gzipStream: false,
+      intakeReq: false,
+      intakeRes: false
+    }
+    let numToComplete = Object.keys(completedFromPart).length
+    const completePart = (part, err) => {
```

Comment:
If I understand the intent of `completePart` -- it's a coordinating function that allows us to update the completed state of the gzip/chopper stream, the request, and the response.  I haven't completely traced out each event's code yet so this might be a naive question -- but is there any risk of a deadlock-ish condition here and is there anything we can/should do about that?

Reply:
1. I guess the comment above "// Handle conclusion ..." isn't close enough to the function to obviously be a description for `completePart()`? If so, I'll move that comment block to be more obvious.
2. I don't believe there is a deadlock, no. There is always the `serverTimeout`, which starts a (default 30s) clock on idleness on the socket. As the comment further above states: `This is a catch-all fallback for an otherwised wedged connection.`.

Reply:
The name and comment are fine -- _If I understand the intent_ is more about me than this code ;)

And 👍 on the no-deadlocks reasoning, resolving this comment.